### PR TITLE
Implement Redacted Send/Receive

### PR DESCRIPTION
**Pull request by pcd1193182**
This PR contains the implementation of Redacted Send and Receive. This is a new feature for ZFS that allows users to selectively not back up parts of their data for space savings, security/privacy concerns, or other reasons. It also contains a number of related features, including enhancements to zdb and zstreamdump, an improvement to the ASSERT and VERIFY macros, the ability to do send size estimation from bookmarks, a more accurate send size estimation process, type support for featureflags, and new tests for all the various new features. There is also a substantial refactor of the zfs send/recv logic to make it more readable and maintainable. Finally, additions have been made to the documentation to reflect the new features and logic.

### Motivation and Context
Redacted send/receive allows users to send subsets of their data to a target system. One possible use case for this feature is to not transmit sensitive information to a data warehousing, test/dev, or analytics environment. Another is to save space by not replicating unimportant data within a given dataset, for example in backup tools like [zrepl](https://www.youtube.com/watch?v=U4TUPQzZzPk&index=4&list=PLaUVvul17xSe0pC6sCirlZXYqICP09Y8z&t=0s).

### Description
Redacted send/receive is a three-stage process. First, a clone (or clones) is made of the snapshot to be sent to the target. In this clone (or clones), all unnecessary or unwanted data is removed or modified. This clone is then snapshotted to create the "redaction snapshot" (or snapshots). Second, the new `zfs redact` command is used to create a redaction bookmark. The redaction bookmark stores the list of blocks in a snapshot that were modified by the redaction snapshot(s). Finally, the redaction bookmark is passed as a parameter to zfs send. When sending to the snapshot that was redacted, the redaction bookmark is used to filter out blocks that contain sensitive or unwanted information, and those blocks are not included in the send stream. When sending from the redaction bookmark, the blocks it contains are considered as candidate blocks in addition to those blocks in the destination snapshot that were modified since the creation_txg of the redaction bookmark. This step is necessary to allow the target to rehydrate data in the case where some blocks are accidentally or unnecessarily modified in the redaction snapshot.

The changes to bookmarks to enable fast space estimation involve adding deadlists to bookmarks. There is also logic to manage the life cycles of these deadlists.

The new size estimation process operates in cases where previously an accurate estimate could not be provided. In those cases, a send is performed where no data blocks are read, reducing the runtime significantly and providing a byte-accurate size estimate.

To make part of this diff easier to review, I have created a gist that contains a diff between the bottom half of the old dmu_send.c and the current dmu_recv.c. This should make it easier to compare the content changes that occurred while the refactor was occurring. That gist is located [here](https://gist.github.com/pcd1193182/818434499d91586280cdb5349a0c683e).

### How Has This Been Tested?
Many new tests have been added for all the changes in this pull request.  In addition, the changes have been running in production at Delphix for months or years, and are considered reliable.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: include/spl/sys/debug.h [hunk 1]
```diff
@@ -107,13 +144,13 @@ void spl_dumpstack(void);
...
 #else
 
-#define	ASSERT(cond)		VERIFY(cond)
+#define	ASSERT3B VERIFY3B
+#define	ASSERT3S VERIFY3S
+#define	ASSERT3U VERIFY3U
+#define	ASSERT3P VERIFY3P
+#define	ASSERT0 VERIFY0
+#define	ASSERT VERIFY
 #define	ASSERTV(x)		x
```
Comment (behlendorf) [inline include/spl/sys/debug.h]: Food for thought, we could retire `ASSERTV`.  In nearly all instances it could be replaced by `__attribute__ ((unused))` which is understood by both `gcc` and `llvm`.

File: cmd/zfs/zfs_main.c [hunk 2]
```diff
@@ -3941,6 +4027,9 @@ zfs_do_snapshot(int argc, char **argv)
 	return (-1);
 }
 
+
+#define	REDACT_OPT	1024
```
[the remark this answers is not in the document]
Comment (pcd1193182) [reply cmd/zfs/zfs_main.c]: Mostly I couldn't think of a good letter. R, e, d, c, and t are all taken. Could do -a, but that usually means something like "all". Could do -r, but that's usually a variant of a recursive option. That, and this isn't going to be a heavily used flag, so it didn't feel mandatory to have a short option for it.

File: module/zfs/dsl_bookmark.c [hunk 3]
```diff
@@ -459,3 +950,559 @@ dsl_bookmark_destroy(nvlist_t *bmarks, nvlist_t *errors)
...
+		 * blkid are not set to UINT64_MAX.
+		 */
+		return (SET_ERROR(EINVAL));
+	}
+
+	/*
+	 * Binary search for the point to resume from.  The goal is to minimize
+	 * the number of disk reads we have to perform.
+	 */
+	buf = kmem_alloc(bufsize, KM_SLEEP);
```
[the remark this answers is not in the document]
Comment (pcd1193182) [reply module/zfs/dsl_bookmark.c]: The redaction list right now only contains 128k blocks, but that is not set in stone, it could be increased (or decreased) later for performance reasons.

File: cmd/zdb/zdb.c [hunk 4]
```diff
@@ -6212,10 +6431,11 @@ main(int argc, char **argv)
 			free(checkpoint_target);
 	}
 
-	if (os != NULL)
+	if (os != NULL) {
```
Comment (Ramzec) [inline cmd/zdb/zdb.c]: As I remember {} are not required if the body is single-line, so why did you change this condition-block?

File: cmd/zdb/zdb.c [hunk 5]
```diff
@@ -1781,6 +1822,126 @@ dump_full_bpobj(bpobj_t *bpo, const char *name, int indent)
...
+dump_bookmarks(objset_t *os, const char *osname, int verbosity)
+{
+	zap_cursor_t zc;
+	zap_attribute_t attr;
+	dsl_dataset_t *ds = dmu_objset_ds(os);
+	dsl_pool_t *dp = spa_get_dsl(os->os_spa);
+	objset_t *mos = os->os_spa->spa_meta_objset;
+	if (verbosity < 4)
+		return;
+	VERIFY0(dsl_pool_hold(osname, FTAG, &dp));
```
[the remark this answers is not in the document]
Comment (Ramzec) [reply cmd/zdb/zdb.c]: I see dump_bookmak() is called inside of block (lines 6359-6372) "dsl_pool_hold() / dsl_pool_rele()". So this dsl_pool_hold() is not required at all. In this case you need to add an VERIFY() that the pool is held.

[the remark this answers is not in the document]
Comment (pcd1193182) [reply cmd/zdb/zdb.c]: In the second comment, you appear to be confusing dump_bookmark() with dump_bookmarks(). This is the latter, and it calls the former.

File: cmd/zdb/zdb.c (hunk 4 above)
Comment (pcd1193182) [reply cmd/zdb/zdb.c]: For consistency with other if() else blocks in the code; see 20 or so lines above, when we have an if () else if() else block where all three clauses are single lines but we use braces anyway.  This change isn't mandatory, but I changed it at an earlier point in the code when this logic got tweaked slightly and didn't remove the change. If it's objectionable, I can revert it.

File: include/sys/dsl_bookmark.h [hunk 6]
```diff
@@ -13,37 +13,113 @@
...
+} dsl_bookmark_node_t;
+
+typedef struct redact_block_phys {
+	uint64_t	rbp_object;
+	uint64_t	rbp_blkid;
+	/*
+	 * The top 16 bits of this field represent the block size in sectors of
+	 * the blocks in question; the bottom 48 bits are used to store the
+	 * number of consecutive blocks that are in the redaction list.  They
+	 * should be accessed using the inline functions below.
```
Comment (ahrens) [inline include/sys/dsl_bookmark.h]: I don't see the functions below; looks like they are in dmu_redact.h.

**Comment by pcd1193182** [comment]
The PR has been rebase on top of master

**Comment by behlendorf** [comment]
@pcd1193182 while I work my way though reviewing and testing this, it looks like the CI caught at least one [memory leak](http://build.zfsonlinux.org/builders/Ubuntu%2017.04%20x86_64%20Coverage%20%28TEST%29/builds/5899/steps/shell_9/logs/kmemleak) which needs to be investigated.  I should be able to get you some feedback by mid next week.

**Comment by kpande** [comment]
> When sending from the redaction bookmark, the blocks it contains are considered as candidate blocks in addition to those blocks in the destination snapshot that were modified since the creation_txg of the redaction bookmark. This step is necessary to allow the target to rehydrate data in the case where some blocks are accidentally or unnecessarily modified in the redaction snapshot.

what exactly does this mean?

**Comment by pcd1193182** [comment]
> what exactly does this mean? how on earth can blocks accidentally or unnecessarily be modified?

The phrasing here could probably be clearer. The idea is that when doing redaction, there is some data that you care about hiding; this is the data you are attempting to redact. However, because of the nature of the data or the surroundings of the data or some quirk of the tools being used, other blocks may be changed when you modify that data  For example, if the data being masked is stored in a database, automatic operations on the database may result in other blocks of the database being modified before a redaction snapshot is taken. If a later version of the masked data doesn't have that automatic operation occur, then those blocks would not be redacted in the later redaction bookmark. In that case, those blocks would be rehydrated when doing an incremental that redacts with that bookmark.

Or it could simply be an accident; if you delete a password in a config file, every block after it will change because the text gets shifted. If you, in a later redaction operation, overwrite the password with garbage of the same length, only that block would be redacted.

**Comment by pcd1193182** [comment]
> @pcd1193182 while I work my way though reviewing and testing this, it looks like the CI caught at least one [memory leak](http://build.zfsonlinux.org/builders/Ubuntu%2017.04%20x86_64%20Coverage%20%28TEST%29/builds/5899/steps/shell_9/logs/kmemleak) which needs to be investigated. I should be able to get you some feedback by mid next week.

Found and fixed the memory leak, will update the PR once I have a passing test run.

**Comment by pcd1193182** [comment]
Working on addressing these test failures now.
